# Electricity tariffs analysis and compliance. 
## 01_Notebook: Dataset download, explore and clean.
---


## Overview.
This project investigates electricity tariffs and revenue protection strategies using data science and machine learning techniques. The primary objectives are to:

- Detect fraudulent activities related to electricity and gas consumption.
- Optimize billing processes for operational efficiency.
- Support regulatory compliance and financial sustainability.
---

## About the dataset.
The dataset originates from the Tunisian Company of Electricity and Gas (STEG), a public non-administrative entity responsible for delivering electricity and gas across Tunisia. STEG has reported significant financial losses—estimated at 200 million Tunisian Dinars—due to widespread meter tampering and fraudulent consumption practices.

## Dataset structure.

The dataset consists of five CSV files, each serving a distinct purpose in the modeling and evaluation pipeline:

- `SampleSubmission.csv`: Template for submitting predictions in a competition format.
- `client_train.csv`: Training data containing client-specific features.
- `client_test.csv`: Test data for evaluating model generalization on unseen clients.
- `invoice_train.csv`: Training data with invoice-level details linked to clients.
- `invoice_test.csv`: Test data for invoice-level prediction tasks.

These files collectively support supervised learning workflows for fraud detection in electricity and gas consumption.

Source: [Fraud detection in electricity and gas consumption on Kaggle](https://www.kaggle.com/datasets/mrmorj/fraud-detection-in-electricity-and-gas-consumption)

---

## Data download & cleaning workflow.

---

### Tec1: Download the dataset.
- Retrieve raw data from source and confirm successful access.

### Tec2: Load and verify the five datasets.
- Read files into dataframes and validate structure and integrity.

### Tec3: Save cleaned raw data locally.
- Store initial cleaned version for reproducibility and backup.

### Tec4: Explore the dataset preliminarily.
- Review shape, types, and basic distributions.

### Tec5: Clean and rename columns for stakeholder localization.
- Standardize column names and align with operational terminology.

### Tec6: Save clean fragmented dataset for export.
- Segment and store final cleaned data for downstream workflows.

### Tec7: Summarise the notebook.
- Conclude the foundational steps for the project.
---

## __Tec1__: Download the dataset.
---

### Install and import necessary packages.

In [3]:
# Install necessary packages
# !pip install kaggle
# !pip install kagglehub

In [4]:
import kagglehub
import pandas as pd
import os


c:\Users\Lenovo\anaconda3\envs\projects_venv\Lib\site-packages\tqdm\auto.py:21: TqdmWarning: IProgress not found. Please update jupyter and ipywidgets. See https://ipywidgets.readthedocs.io/en/stable/user_install.html
  from .autonotebook import tqdm as notebook_tqdm


In [5]:
# Download latest version of the dataset
path = kagglehub.dataset_download("mrmorj/fraud-detection-in-electricity-and-gas-consumption")

# Verify the path
print("Dataset downloaded to:", path)

# Process all CSV files
csv_files = [f for f in os.listdir(path) if f.endswith('.csv')]         # List all CSV files in the directory

def read_messy_csv(file_path):                                          # Function to read messy CSV files
    try:
        df = pd.read_csv(
            file_path, 
            low_memory=False,       # Handle mixed data types
            on_bad_lines='skip',    # Skip problematic lines
            encoding='utf-8'        # Ensure proper encoding
        )
        return df
    except Exception as e:
        print(f"Error reading {file_path}: {e}")
        return None

# Process and print details of each CSV file
for file in csv_files:
    file_path = os.path.join(path, file)
    print(f"\nProcessing {file}:")
    
    df = read_messy_csv(file_path)
    
    if df is not None:
        print(f"Successfully read {file}")
        # print("Columns:", list(df.columns))
        # print("Shape:", df.shape)
        # print("\nFirst few rows:")
        # print(df.head())
        
        # Optional: Save a clean version if needed
        clean_file_path = os.path.join(path, f"clean_{file}")
        df.to_csv(clean_file_path, index=False)
        print(f"Saved clean version to: {clean_file_path}")
    else:
        print(f"Failed to read {file}")

Dataset downloaded to: C:\Users\Lenovo\.cache\kagglehub\datasets\mrmorj\fraud-detection-in-electricity-and-gas-consumption\versions\2

Processing clean_clean_clean_clean_clean_clean_clean_clean_clean_clean_client_test.csv:
Successfully read clean_clean_clean_clean_clean_clean_clean_clean_clean_clean_client_test.csv
Saved clean version to: C:\Users\Lenovo\.cache\kagglehub\datasets\mrmorj\fraud-detection-in-electricity-and-gas-consumption\versions\2\clean_clean_clean_clean_clean_clean_clean_clean_clean_clean_clean_client_test.csv

Processing clean_clean_clean_clean_clean_clean_clean_clean_clean_clean_client_train.csv:
Successfully read clean_clean_clean_clean_clean_clean_clean_clean_clean_clean_client_train.csv
Saved clean version to: C:\Users\Lenovo\.cache\kagglehub\datasets\mrmorj\fraud-detection-in-electricity-and-gas-consumption\versions\2\clean_clean_clean_clean_clean_clean_clean_clean_clean_clean_clean_client_train.csv

Processing clean_clean_clean_clean_clean_clean_clean_clean_cle

## __Tec2__: Load and verify the five datasets.
---

### __First CSV file__: Load and verify. 

In [6]:
# Load the first CSV file to verify.
file1_path = r'C:\Users\Lenovo\.cache\kagglehub\datasets\mrmorj\fraud-detection-in-electricity-and-gas-consumption\versions\2\clean_clean_clean_clean_client_test.csv'
tariffs_df1 = pd.read_csv(file1_path)
tariffs_df1.head()

disrict         client_id  client_catg  region creation_date
0       62     test_Client_0           11     307    28/05/2002
1       69     test_Client_1           11     103    06/08/2009
2       62    test_Client_10           11     310    07/04/2004
3       60   test_Client_100           11     101    08/10/1992
4       62  test_Client_1000           11     301    21/07/1977

### __Second CSV file__: Load and verify.

In [7]:
file2_path = r'C:\Users\Lenovo\.cache\kagglehub\datasets\mrmorj\fraud-detection-in-electricity-and-gas-consumption\versions\2\clean_clean_clean_clean_client_train.csv'
tariffs_df2 = pd.read_csv(file2_path)
tariffs_df2.head()

disrict          client_id  client_catg  region creation_date  target
0       60     train_Client_0           11     101    31/12/1994     0.0
1       69     train_Client_1           11     107    29/05/2002     0.0
2       62    train_Client_10           11     301    13/03/1986     0.0
3       69   train_Client_100           11     105    11/07/1996     0.0
4       62  train_Client_1000           11     303    14/10/2014     0.0

### __Third CSV file__: Load and verify.

In [8]:
file3_path = r'C:\Users\Lenovo\.cache\kagglehub\datasets\mrmorj\fraud-detection-in-electricity-and-gas-consumption\versions\2\clean_clean_clean_clean_invoice_test.csv'
tariffs_df3 = pd.read_csv(file3_path)
tariffs_df3.head()

client_id invoice_date  tarif_type  counter_number  counter_statue  \
0  test_Client_0   2018-03-16          11          651208               0   
1  test_Client_0   2014-03-21          11          651208               0   
2  test_Client_0   2014-07-17          11          651208               0   
3  test_Client_0   2015-07-13          11          651208               0   
4  test_Client_0   2016-07-19          11          651208               0   

   counter_code  reading_remarque  counter_coefficient  consommation_level_1  \
0           203                 8                    1                   755   
1           203                 8                    1                  1067   
2           203                 8                    1                     0   
3           203                 9                    1                   410   
4           203                 9                    1                   412   

   consommation_level_2  consommation_level_3  consommation_level_4  \
0                     0                     0                     0   
1                     0                     0                     0   
2                     0                     0                     0   
3                     0                     0                     0   
4                     0                     0                     0   

   old_index  new_index  months_number counter_type  
0      19145      19900              8         ELEC  
1      13725      14792              8         ELEC  
2      14792      14792              4         ELEC  
3      16122      16532              4         ELEC  
4      17471      17883              4         ELEC

### __Fourth CSV file__: Load and verify.

In [9]:
file4_path = r'C:\Users\Lenovo\.cache\kagglehub\datasets\mrmorj\fraud-detection-in-electricity-and-gas-consumption\versions\2\clean_clean_clean_clean_invoice_train.csv'
tariffs_df4 = pd.read_csv(file4_path)
tariffs_df4.head()

C:\Users\Lenovo\AppData\Local\Temp\ipykernel_23300\1943407127.py:2: DtypeWarning: Columns (4) have mixed types. Specify dtype option on import or set low_memory=False.
  tariffs_df4 = pd.read_csv(file4_path)


client_id invoice_date  tarif_type  counter_number counter_statue  \
0  train_Client_0   2014-03-24          11         1335667              0   
1  train_Client_0   2013-03-29          11         1335667              0   
2  train_Client_0   2015-03-23          11         1335667              0   
3  train_Client_0   2015-07-13          11         1335667              0   
4  train_Client_0   2016-11-17          11         1335667              0   

   counter_code  reading_remarque  counter_coefficient  consommation_level_1  \
0           203                 8                    1                    82   
1           203                 6                    1                  1200   
2           203                 8                    1                   123   
3           207                 8                    1                   102   
4           207                 9                    1                   572   

   consommation_level_2  consommation_level_3  consommation_level_4  \
0                     0                     0                     0   
1                   184                     0                     0   
2                     0                     0                     0   
3                     0                     0                     0   
4                     0                     0                     0   

   old_index  new_index  months_number counter_type  
0      14302      14384              4         ELEC  
1      12294      13678              4         ELEC  
2      14624      14747              4         ELEC  
3      14747      14849              4         ELEC  
4      15066      15638             12         ELEC

### __Fifth CSV file__: Load and verify.

In [10]:
file5_path = r'C:\Users\Lenovo\.cache\kagglehub\datasets\mrmorj\fraud-detection-in-electricity-and-gas-consumption\versions\2\clean_clean_clean_clean_SampleSubmission (2).csv'
tariffs_df5 = pd.read_csv(file5_path)
tariffs_df5.head()

client_id    target
0     test_Client_0  0.957281
1     test_Client_1  0.996425
2    test_Client_10  0.612359
3   test_Client_100  0.776933
4  test_Client_1000  0.571046

---

## __Tec3__: Save cleaned raw data locally.

Although the raw dataset was downloaded using `kagglehub` to minimize local storage usage, save a cleaned version of each file for reproducibility and modular access.

- **Directory**: `data/raw/`
- **Files**: `clean_client_train.csv`, `clean_client_test.csv`, `clean_invoice_train.csv`, `clean_invoice_test.csv`, `clean_SampleSubmission.csv`

---

In [11]:
# Define the raw data directory and ensure it exists.
raw_path = r"C:\Users\Lenovo\OneDrive\Desktop\4IR_DataScience\DataScienceEnvironment\my_projects\electricity_tariffs_revenue_protection\electricity_tariffs_analysis_compliance\data\raw"
os.makedirs(raw_path, exist_ok=True)

# Save cleaned raw files (assuming file1_path to file5_path are defined).
tariffs_df1.to_csv(f"{raw_path}/clean_client_train.csv", index=False)
tariffs_df2.to_csv(f"{raw_path}/clean_client_test.csv", index=False)
tariffs_df3.to_csv(f"{raw_path}/clean_invoice_train.csv", index=False)
tariffs_df4.to_csv(f"{raw_path}/clean_invoice_test.csv", index=False)
tariffs_df5.to_csv(f"{raw_path}/clean_SampleSubmission.csv", index=False)


## __Tec4__: Explore the dataset preliminarily.
---

In [12]:
# Preliminary exploration of the dataset
dataset = [tariffs_df1, tariffs_df2, tariffs_df3, tariffs_df4, tariffs_df5]

# Explore the respective DataFrames of the dataset

for i, df in enumerate(dataset):
    print(f'DataFrame {i+1} info:')
    print(df.info())
    print('Dimensions:', df.shape)
    print('\nMissing values:\n', df.isnull().sum())
    print('\nStats discription:\n', df.describe())
    print('\n')

DataFrame 1 info:
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 58069 entries, 0 to 58068
Data columns (total 5 columns):
 #   Column         Non-Null Count  Dtype 
---  ------         --------------  ----- 
 0   disrict        58069 non-null  int64 
 1   client_id      58069 non-null  object
 2   client_catg    58069 non-null  int64 
 3   region         58069 non-null  int64 
 4   creation_date  58069 non-null  object
dtypes: int64(3), object(2)
memory usage: 2.2+ MB
None
Dimensions: (58069, 5)

Missing values:
 disrict          0
client_id        0
client_catg      0
region           0
creation_date    0
dtype: int64

Stats discription:
             disrict   client_catg        region
count  58069.000000  58069.000000  58069.000000
mean      63.510617     11.507224    206.018461
std        3.356708      4.395923    104.144028
min       60.000000     11.000000    101.000000
25%       62.000000     11.000000    103.000000
50%       62.000000     11.000000    107.000000
75%       69

## Dataset overview.
---
### File 1 was loaded into __tariffs_df1__ dataframe.
- **Columns**: ['district', 'client_id', 'client_catg', 'region', 'creation_date']
- **Shape**: (58069, 5)
- **Missing Values**: 
zero missing values

### File 2 was loaded into __tariffs_df2__ dataframe.
- **Columns**: ['disrict', 'client_id', 'client_catg', 'region', 'creation_date', 'target']
- **Shape**: (135493, 6)
- **Missing Values**: 
zero missing values

### File 3 was loaded into __tariffs_df3__ dataframe.
- **Columns**: ['client_id', 'invoice_date', 'tariff_type', 'meter_number', 'meter_status', 'meter_code', 'reading_remark', 'meter_coefficient', 'consumption_level_1', 'consumption_level_2', 'consumption_level_3', 'consumption_level_4', 'old_reading', 'new_reading', 'number_months', 'meter_type']
- **Shape**: (1939730, 16)
- **Missing Values**: 
  zero missing values

### File 4 was loaded into __tariffs_df4__ dataframe.
- **Columns**: ['client_id', 'invoice_date', 'tariff_type', 'meter_number', 'meter_status', 'meter_code', 'reading_remark', 'meter_coefficient', 'consumption_level_1', 'consumption_level_2', 'consumption_level_3', 'consumption_level_4', 'old_reading', 'new_reading', 'number_months', 'meter_type']
- **Shape**: (4476749, 16)
- **Missing Values**: 
zero missing values

### File 5 was loaded into __tariffs_df5__ dataframe.
- **Columns**: ['client_id', 'target']
- **Shape**: (58069, 2)
- **Missing Values**: 
zero mising values

## __Tec5__: Clean and rename columns for stakeholder localization.
---
The original datasets used in this analysis were sourced from Kaggle and originate from a Francophone country, resulting in column names that were predominantly French-inclined.

To enhance clarity and ensure stakeholder familiarity, all column names were systematically renamed to reflect terminology aligned with the local context and language of the targeted stakeholders. This localization effort supports:

- Improved readability for non-French-speaking analysts and reviewers.
- Seamless integration into stakeholder-facing documentation and dashboards.
- Reduced cognitive friction during model interpretation and compliance review.

> This renaming step ensures that technical outputs remain accessible, interpretable, and aligned with operational language norms.
---


In [13]:
# foward fill missing values
tariffs_df1.ffill(inplace=True)
tariffs_df2.ffill(inplace=True)
tariffs_df3.ffill(inplace=True)
tariffs_df4.ffill(inplace=True)
tariffs_df5.ffill(inplace=True)

# Remove duplicates
tariffs_df1.drop_duplicates(inplace=True)
tariffs_df2.drop_duplicates(inplace=True)
tariffs_df3.drop_duplicates(inplace=True)
tariffs_df4.drop_duplicates(inplace=True)
tariffs_df5.drop_duplicates(inplace=True)

# Rename columns in tariffs_df1 and tariffs_df2.
tariffs_df1.rename(columns={'disrict': 'district'}, inplace=True)
tariffs_df2.rename(columns={'disrict': 'district'}, inplace=True)

# Rename columns in tariffs_df3.
tariffs_df3.rename(columns={
    'counter_number': 'meter_number',
    'counter_statue': 'meter_status',
    'counter_code': 'meter_code',
    'tarif_type': 'tariff_type',
    'reading_remarque': 'reading_remark',
    'counter_coefficient': 'meter_coefficient',
    'consommation_level_1': 'consumption_level_1',
    'consommation_level_2': 'consumption_level_2',
    'consommation_level_3': 'consumption_level_3',
    'consommation_level_4': 'consumption_level_4',
    'old_index': 'old_reading',
    'months_number': 'number_months',
    'new_index': 'new_reading',
    'counter_type': 'meter_type'
}, inplace=True)

# Rename columns in tariffs_df4.
tariffs_df4.rename(columns={
    'counter_number': 'meter_number',
    'counter_statue': 'meter_status',
    'counter_code': 'meter_code',
    'tarif_type': 'tariff_type',
    'reading_remarque': 'reading_remark',
    'counter_coefficient': 'meter_coefficient',
    'consommation_level_1': 'consumption_level_1',
    'consommation_level_2': 'consumption_level_2',
    'consommation_level_3': 'consumption_level_3',
    'consommation_level_4': 'consumption_level_4',
    'old_index': 'old_reading',
    'months_number': 'number_months',
    'new_index': 'new_reading',
    'counter_type': 'meter_type'
}, inplace=True)


### Inspect the cleaned dataframes.

In [14]:
# inspect the cleaned .ataframes.
for i, df in enumerate(dataset):
    print(f'DataFrame {i+1} after cleaning:')
    print(f'Columns: {list(df.columns)}')
    print(df.head())
    print('Dimensions:', df.shape)
    print('\nMissing values:\n', df.isnull().sum())
    # print('\nStats description:\n', df.describe())
    print('\n')

DataFrame 1 after cleaning:
Columns: ['district', 'client_id', 'client_catg', 'region', 'creation_date']
   district         client_id  client_catg  region creation_date
0        62     test_Client_0           11     307    28/05/2002
1        69     test_Client_1           11     103    06/08/2009
2        62    test_Client_10           11     310    07/04/2004
3        60   test_Client_100           11     101    08/10/1992
4        62  test_Client_1000           11     301    21/07/1977
Dimensions: (58069, 5)

Missing values:
 district         0
client_id        0
client_catg      0
region           0
creation_date    0
dtype: int64


DataFrame 2 after cleaning:
Columns: ['district', 'client_id', 'client_catg', 'region', 'creation_date', 'target']
   district          client_id  client_catg  region creation_date  target
0        60     train_Client_0           11     101    31/12/1994     0.0
1        69     train_Client_1           11     107    29/05/2002     0.0
2        62    tr

### Data cleaning and column renaming summary.

#### Dataframes: `tariffs_df1` to `tariffs_df5`.

**General gleaning steps:**
- **Missing values**: Applied forward fill using `fillna(method='ffill')` to preserve temporal continuity.
- **Duplicates**: Removed redundant entries with `drop_duplicates()` to ensure data integrity.

---

#### Column renaming for localization and clarity.

Renaming was performed to align with South African electricity terminology and improve stakeholder comprehension.

#### Common renaming across `tariffs_df3` and `tariffs_df4`:
| Original column           | Renamed column         |
|---------------------------|------------------------|
| `counter_number`          | `meter_number`         |
| `counter_statue`          | `meter_status`         |
| `counter_code`            | `meter_code`           |
| `reading_remarque`        | `reading_remark`       |
| `counter_coefficient`     | `meter_coefficient`    |
| `consommation_level_1`    | `consumption_level_1`  |
| `consommation_level_2`    | `consumption_level_2`  |
| `consommation_level_3`    | `consumption_level_3`  |
| `consommation_level_4`    | `consumption_level_4`  |
| `old_index`               | `old_reading`          |
| `new_index`               | `new_reading`          |
| `months_number`           | `number_months`        |
| `counter_type`            | `meter_type`           |

#### Specific renaming:
- **`tariffs_df1` and `tariffs_df2`**:
  - `disrict` → `district`

---

## Feature description development & documentation

Since the dataset lacks predefined column descriptions, we developed a custom schema grounded in domain knowledge of electricity billing and fraud detection. This enhances interpretability, supports reproducibility, and aligns with stakeholder narratives.

### Methodology
- **Analyze column names**: Inferred meanings using electricity billing and fraud detection terminology.
- **Contextual mapping**: Linked features to operational roles and fraud risk indicators.
- **Documentation**: Provided concise, actionable descriptions for each column across all dataframes.

---

### tariff_df1
| Column         | Description |
|----------------|-------------|
| `district`     | Client’s district; useful for identifying geographical fraud patterns. |
| `client_id`    | Unique client identifier; essential for tracking and anomaly detection. |
| `client_catg`  | Client category (e.g., residential, commercial); informs consumption norms. |
| `region`       | Broader geographical region; complements district-level analysis. |
| `creation_date`| Account creation date; helps assess client tenure and lifecycle behavior. |

---

### tariff_df2
Includes all columns from `tariff_df1`, plus:
| Column               | Description |
|----------------------|-------------|
| `target`             | Binary fraud indicator (0 = no fraud, 1 = suspected fraud); modeling target. |

---

### tariff_df3
| Column               | Description |
|----------------------|-------------|
| `invoice_date`       | Date of invoice issuance; tracks billing cycles. |
| `tarif_type`         | Type of tariff applied; different tariffs may carry distinct fraud risks. |
| `counter_number`     | Meter number; links consumption to physical infrastructure. |
| `counter_statue`     | Meter status (e.g., active/inactive); inactive meters may signal fraud. |
| `counter_code`       | Meter code; used for verification. |
| `reading_remarque`   | Notes on meter readings; may explain anomalies. |
| `counter_coefficient`| Coefficient affecting consumption calculation. |
| `consommation_level_1` to `level_4` | Tiered consumption levels; useful for behavioral segmentation. |
| `old_index`          | Previous meter reading; used in consumption calculation. |
| `new_index`          | Current meter reading; same as above. |
| `months_number`      | Time span between readings; affects consumption normalization. |
| `counter_type`       | Meter type; may influence fraud risk. |

---

### tariff_df4
Same structure and columns as `tariff_df3`.

---

### tariff_df5
| Column      | Description |
|-------------|-------------|
| `client_id` | Same as above. |
| `target`    | Same as `tariff_df2`; used for supervised fraud modeling. |

---

### Applying the results to the City of Tshwane
This schema enables localized fraud analysis by:
- Identifying high-risk regions and client categories.
- Linking consumption anomalies to meter status and coefficients.
- Using the `target` column to train models and validate operational strategies.

---



## __Tec6__: Save clean fragmented dataset for export.
---

To maintain a modular and scalable directory structure, the cleaned fragmented dataset is saved in a dedicated sub-folder:

- **Directory**: `data/processed/fragmented_data/`

- **Filename**: `clean_fragmented_data.csv`

This approach avoids congestion in the main `processed` folder and supports thematic organization of intermediate outputs for future reuse and repackaging.


In [15]:
import os

# Create sub-folders under processed
clean_fragmented_data_path = r'C:\Users\Lenovo\OneDrive\Desktop\4IR_DataScience\DataScienceEnvironment\my_projects\electricity_tariffs_revenue_protection\electricity_tariffs_analysis_compliance\data\processed\fragmented_data'
os.makedirs(clean_fragmented_data_path, exist_ok=True)

# Save tariff-related DataFrames
tariffs_df1.to_csv(f"{clean_fragmented_data_path}/tariffs_df1.csv", index=False)
tariffs_df2.to_csv(f"{clean_fragmented_data_path}/tariffs_df2.csv", index=False)
tariffs_df3.to_csv(f"{clean_fragmented_data_path}/tariffs_df3.csv", index=False)
tariffs_df4.to_csv(f"{clean_fragmented_data_path}/tariffs_df4.csv", index=False)
tariffs_df5.to_csv(f"{clean_fragmented_data_path}/tariffs_df5.csv", index=False)

print("Tariff datasets saved to 'data/processed/fragmented_data/'")


Tariff datasets saved to 'data/processed/fragmented_data/'


## __Tec7__ Summarise the notebook: Data download, exploration, and cleaning.
---
This notebook established the foundational steps for the electricity tariffs and revenue protection analysis project. Key milestones include:

- **Dataset acquisition**: Downloaded raw data using `kagglehub` for efficient, on-demand access.
- **Preliminary exploration**: Assessed structure, missing values, and variable distributions across five core files.
- **Localization and renaming**: Applied domain-specific terminology to enhance stakeholder comprehension and modeling relevance.
- **Feature documentation**: Developed a custom schema to describe each column’s analytical role, supporting transparency and reproducibility.
- **Data export**: Saved cleaned datasets into modular sub-folders under `data/processed/fragmented_data`.

---

## Next steps: `data_packaging.ipynb`

---